File: Java/Tests/jdi-uitest-wintests/src/test/java/com/epam/jdi/uitests/testing/unittests/tests/complex/DropdownExpandedTests.java
package com.epam.jdi.uitests.testing.unittests.tests.complex;

import com.epam.jdi.uitests.core.interfaces.complex.IDropDown;
import com.epam.jdi.uitests.testing.unittests.InitTests;
import com.epam.jdi.uitests.testing.unittests.enums.Colors;
import com.epam.web.matcher.testng.Check;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

import java.io.IOException;
import java.lang.reflect.Method;
import java.util.List;

import static com.epam.jdi.uitests.core.preconditions.PreconditionsState.isInState;
import static com.epam.jdi.uitests.testing.unittests.enums.Colors.Blue;
import static com.epam.jdi.uitests.testing.unittests.enums.Colors.Colors;
import static com.epam.jdi.uitests.testing.unittests.enums.Preconditions.METALS_AND_COLORS_PAGE;
import static com.epam.jdi.uitests.testing.unittests.pageobjects.EpamJDIApplication.metalsColorsPage;
import static com.epam.jdi.uitests.testing.unittests.tests.complex.CommonActionsData.*;
import static com.epam.web.matcher.testng.Assert.areEquals;
import static com.epam.web.matcher.testng.Assert.listEquals;
import static java.util.Arrays.asList;

/**
 * Created by Roman_Iovlev on 9/15/2015.
 */
public class DropdownExpandedTests extends InitTests {
    private static final List<String> oddOptions = asList("Colors", "Red", "Green", "Blue", "Yellow");

    private IDropDown<Colors> colors() {
        return metalsColorsPage.colors;
    }

    @BeforeMethod
    public void before(Method method) throws IOException {
        isInState(METALS_AND_COLORS_PAGE, method);
        colors().expand();
    }

    @Test
    public void selectStringTest() {
        colors().select("Blue");
        checkAction("Colors: value changed to Blue");
    }

    @Test
    public void selectIndexTest() {
        colors().select(4);
        checkAction("Colors: value changed to Blue");
    }

    @Test
    public void selectEnumTest() {
        colors().select(Blue);
        checkAction("Colors: value changed to Blue");
    }

    @Test
    public void getOptionsTest() {
        listEquals(colors().getOptions(), oddOptions);
    }

    @Test
    public void getNamesTest() {
        listEquals(colors().getNames(), oddOptions);
    }

    @Test
    public void getValuesTest() {
        listEquals(colors().getValues(), oddOptions);
    }

    @Test
    public void getOptionsAsTextTest() {
        areEquals(colors().getOptionsAsText(), "Colors, Red, Green, Blue, Yellow");
    }

    @Test
    public void setValueTest() {
        colors().setValue("Blue");
        checkAction("Colors: value changed to Blue");
    }

    @Test
    public void getNameTest() {
        areEquals(colors().getName(), "Colors");
    }

    // Fails
    @Test
    public void getSelectedTest() {
        areEquals(colors().getSelected(), "Colors");
    }

    @Test
    public void getSelectedIndexTest() {
        checkActionThrowError(() -> colors().getSelectedIndex(), noElementsMessage); // isDisplayed not defined
    }

    @Test
    public void isSelectedTest() {
        areEquals(colors().isSelected("Colors"), true);
    }

    @Test
    public void isSelectedEnumTest() {
        areEquals(colors().isSelected(Colors), true);
    }

    @Test
    public void waitSelectedTest() {
        new Check("WaitSelected")
                .hasNoExceptions(() -> colors().waitSelected("Colors"));
    }

    @Test
    public void waitSelectedEnumTest() {
        new Check("WaitSelected")
                .hasNoExceptions(() -> colors().waitSelected(Colors));
    }

    @Test
    public void getValueTest() {
        areEquals(colors().getValue(), "Colors");
    }
}

File: Python/Test/jdi_uitests_webtests/test/complex/dropdown_expanded_tests.py
from JDI.jdi_assert.testing.assertion import Assert
from Test.jdi_uitests_webtests.main.enums.entities import Colors
from Test.jdi_uitests_webtests.main.enums.preconditions import Preconditions
from Test.jdi_uitests_webtests.main.page_objects.epam_jdi_site import EpamJDISite
from Test.jdi_uitests_webtests.main.utils.common_action_data import CommonActionsData
from Test.jdi_uitests_webtests.test.init_tests import InitTests


class DropdownExpandedTests(InitTests):

    odd_options = ["Colors", "Red", "Green", "Blue", "Yellow"]

    dropdown = EpamJDISite.metals_colors_page.color_dropdown

    def setUp(self):
        Preconditions.METALS_AND_COLORS_PAGE.is_in_state()
        self.dropdown.expand()

    def test_select_string(self):
        self.dropdown.select("Blue")
        CommonActionsData.check_action("Colors: value changed to Blue")

    def test_select_index(self):
        self.dropdown.select(4)
        CommonActionsData.check_action("Colors: value changed to Blue")

    def test_select_enum(self):
        self.dropdown.select(Colors.BLUE)
        CommonActionsData.check_action("Colors: value changed to Blue")

    def test_get_options(self):
        Assert.assert_equal(self.dropdown.get_options(), self.odd_options)

    def test_get_names(self):
        Assert.assert_equal(self.dropdown.get_names(), self.odd_options)

    def test_get_values(self):
        Assert.assert_equal(self.dropdown.get_values(), self.odd_options)

    def test_get_options_as_text(self):
        Assert.assert_equal(self.dropdown.get_options_as_text(), "Colors, Red, Green, Blue, Yellow")

    def test_set_value(self):
        self.dropdown.set_value("Blue")
        CommonActionsData.check_action("Colors: value changed to Blue")

    def test_get_name(self):
        Assert.assert_equal(self.dropdown.get_name(), "color_dropdown")

    def test_get_selected(self):
        Assert.assert_equal(self.dropdown.get_selected(), "Colors")

    def test_get_selected_index(self):
        CommonActionsData.check_action_throw_error(lambda: self.dropdown.get_selected_index(), CommonActionsData.no_elements_message())

    def test_is_selected(self):
        Assert.assert_equal(self.dropdown.is_selected("Colors"), True)

    def test_is_selected_enum(self):
        Assert.assert_equal(self.dropdown.is_selected(Colors.COLORS), True)

    def test_get_value(self):
        Assert.assert_equal(self.dropdown.get_value(), "Colors")